### PR TITLE
buffer: Don't lock buffers when only reading state or params.

### PR DESCRIPTION
**Pull request by akloniex**
There was an overkill with components locking buffers each time they
read r/w pointers or free/avail, while simply invalidating the buffer's
control block would suffice.
By changing this behaviour we greatly reduce amount of needlessly
acquired locks.

Signed-off-by: Artur Kloniecki <arturx.kloniecki@linux.intel.com>

### CODE REVIEW

**Comment by akloniex** [comment]
@paulstelian97 the reading core would not cause cache to writeback, rather only invalidate. There's no reason for it to writeback.
@lgirdwood This is to eliminate unnecessary locking and cache writeback (which is done as part of buffer_unlock routine), when it is not strictly necessary. All state changes, be it r/w pointers update, free/avail recalculation, params update, are guarded by locks, either preceding params change or as part of buffer produce/consume/reset routines.
As it is right now on master, each buffer_lock call, attempts to acquire a lock, and invalidates cache after lock is acquired. Each buffer_unlock call does wtb and inv of the memory and then releases the lock. This ensures that all state changes are coherent. Although, same procedure was used to read the state, while just cache invalidation alone would suffice.
The only 'risk' in such case is that concurrent cores might under-produce or under-consume, having slightly underestimated free/avail read from the cache, but that's alright, as in next cycle they would make up for the 'slack'. And having reads lockless presents benefits of lesser complexity, cleaner code and having less synchronization points in runtime.

**Comment by paulstelian97** [comment]
Is it not possible that a "non-atomic" writeback will also occur due to the respective cache line being evicted for some weird reason in certain usage patterns? Also, if the other core is doing a writeback (whether intended or not) WHILE you are reading the data here, what is happening? Are we guaranteed to not see partially written data in ways that would bother us? (i.e. the 32-bit words themselves would be read atomically at least).

My prior comment is possibly mistakenly assuming that Xtensa is working on a fully coherent cache (it doesn't?).

I do see the point of your PR, but I want at least comprehensive testing on all platforms which have multiple Xtensa cores to ensure this doesn't introduce any _very subtle_ issue. I still have flashbacks from when I fixed that edge case in the memory allocator back in December, I don't want this to happen here (and SMP is even worse because it won't even be as reproducible, while on that allocator bug I had a reproducible case that pointed me in the wrong place to search for the bug)

**Comment by lgirdwood** [comment]
@akloniex can you give us example usage here in a component copy(), just a mix of the C APIs and psuedo code.

**Comment by akloniex** [comment]
@lgirdwood 
Component copy() flow should look something like this:
``` C
static int component_copy(struct comp_dev *dev)
{
    /* fetch connected buffers */
    struct comp_buffer *source = ....
    struct comp_buffer *sink = ....

    /* invalidate control blocks of the buffers in order to properly read
     * the free and avail values */
    buffer_control_invalidate(source);
    buffer_control_invalidate(sink);

    /* calculate number of frames that will be available to process and fit
     * into the sink, as well as data sizes */
    uint32_t frames = audio_stream_avail_frames(&source->stream, &sink->stream);
    uint32_t source_bytes = frames * audio_stream_frame_bytes(&source->stream);
    uint32_t sink_bytes = frames * audio_stream_frame_bytes(&sink->stream);

    /* invalidate required amount of data from the source buffer in order to
     * properly read the data */
    buffer_invalidate(source, source_bytes);

    /* perform the processing of the data and put the results into the sink */
    _component_processing_function_wrapper_(source, sink, frames)

    /* writeback the sink buffer data */
    buffer_writeback(sink, sink_bytes);

    /* perform state update of the buffers */
    comp_update_buffer_produce(sink, sink_bytes);
    comp_update_buffer_consume(source, source_bytes);

    return 0;
}
```

`buffer_control_invalidate()`, `buffer_invalidate()` and `buffer_writeback()` have an effect iff target buffer is `inter_core`, as that's the only case in which any of these have a meaning.
`comp_update_buffer_produce()` and `comp_update_buffer_consume()` contain lock guards incorporated within them, so that the state change is done within critical section.

**Comment by lgirdwood** [comment]
Thanks @akloniex - this is worthy of a sof-docs technical note as part of the component development docs. Can you add update the sof-docs to reflect the what we merge due to this PR.

I think I would like to simplify the usage somewhat in that we should do the buffer control invalidate for sink and source automatically within audio_stream_avail_frames() i.e.
``` C
static int component_copy(struct comp_dev *dev)
{
    /* fetch connected buffers */
    struct comp_buffer *source = ....
    struct comp_buffer *sink = ....

    /*  invalidate buffer_control here and source frames */
    uint32_t frames = audio_stream_avail_frames(&source->stream, &sink->stream);

   /* calculate number of frames that will be available to process and fit
     * into the sink, as well as data sizes */
    uint32_t source_bytes = frames * audio_stream_frame_bytes(&source->stream);
    uint32_t sink_bytes = frames * audio_stream_frame_bytes(&sink->stream);

    /* perform the processing of the data and put the results into the sink */
    _component_processing_function_wrapper_(source, sink, frames);

    /* perform state update of the buffers  - calls produce and consume - does writeback of control data and sink data */
    audio_stream_commit_frames(source,sink, source_bytes, sink_bytes);

    return 0;
}
```
Please see that I've merged some of the calls together and have created a new call ```audio_stream_commit_frames(source,sink, source_bytes, sink_bytes);```. This make the usage now very simple as developers just have 1 calll beofre and 1 call after processing data.

**Comment by akloniex** [comment]
@lgirdwood That a good idea, but I'm afraid that it might be suboptimal for components with multiple sources/sinks. For SIMO, source would be invalidated for each output and same for the sink in MISO case. Obviously, for such components, developers could use the lower-level control invalidation manually, but that would introduce inconsistency across different types of components.
For the commit function things get slightly worse, as repeated invalidation wouldn't cause that much trouble, just performance decrease, but committing multiple times on same buffer is just wrong.
We could go another level deeper with the abstraction, and pass arrays of sinks and buffers, to account for these cases, but that would make the code hard to read and write, and I don't know if we could properly account for some specific cases, as I'm not aware of all the specific cases that might come in the future.

So to sum up:
- I can aggregate invalidation calls and add invalidation to the `audio_stream_avail_frames()` - that would require moving locks from comp_buffer to audio_stream, and for non-SISO components would either cause multiple invalidations or force lower-level API usage
- I can aggregate produce/consume calls into commit_frames - but not on audio_stream level, but rather on buffer level, due to consume/produce notifications that buffers broadcast. This also forces lower-level API usage for non-SISO components.

After we resolve these concerns I'll update sof-docs.

**Comment by lgirdwood** [comment]
@akloniex ok, I'm just looking for constency here with the APIs given the usage should be simple and balanced. i.e. we did not have a direct buffer_control_writeback. It sound like you have some ideas for higher level API that are both simple (fewer calls) and balanced for both MISO and SIMO alongside SISO processing.

**Comment by paulstelian97** [comment]
@akloniex If this PR is still useful in any fashion, please re-send it based on the "main" branch.
